README.md
# app_solr

A cookbook that install Solr with basic settings. Only tested with Solr versions 5.x to 7.x.

## Supported Platforms

LTS version of Ubuntu >= 20.04

## Resources

### app_solr_standalone

Install a local standalone (not cluster) Solr.

```ruby
app_solr_standalone '7.7.3' do
  solr_java_mem '-Xms512m -Xmx512m'
end
```

#### Actions

- `install` - Install Solr locally in standalone mode

#### Properties

- `version` - Solr version to install. Default to name of resource.
- `apt_packages` - OS packages to install. Default: _openjdk-11_ packages for Solr 7.x and _openjdk-8_ packages for Solr 6.x and below.
- `extract_dir` - Path to which tarball will be extracted. Default: `/opt`.
- `solr_user` - User to run Solr. Will be auto-created by installer. Default: `solr`.
- `set_ulimits` - If true, :solr_user ulimits will be maximized to 65535. Default: true.
- `solr_dir` - Main Solr directory. Default: `/var/solr`.
- `force_install` - Set to true when upgrading Solr. Default: false.

Settings that will be included directly into `/etc/default/solr.in.sh`:
- `solr_home`
- `solr_pid_dir`
- `solr_logs_dir`
- `solr_host`
- `solr_port`
- `solr_java_mem`
- `log4j_props`

### app_solr_core

Create a Solr core.

```ruby
app_solr_core 'myapp1' do
  use_custom_solrconfig true
  solrconfig_source     'solrconfig.xml'
  solrconfig_cookbook   'mycookbook'
end
```

#### Actions

- `create` - Create the core

#### Properties

- `name` - Name of core. Defaults to name of resource.
- `extract_dir` - Path to which Solr installer tarball was extracted. Default: `/opt`.
- `solr_user` - User that runs Solr. Default: `solr`.
- `solr_home` - Location of Solr home relative to :solr_dir. Default: `data`.
- `use_custom_solrconfig` - Set to true in order to use custom `solrconfig.xml` file. Default: false.
- `solrconfig_source` - cookbook_file source for custom `solrconfig.xml` file.
- `solrconfig_cookbook` - cookbook_file cookbook for custom `solrconfig.xml` file.
- `use_custom_schema` - Set to true when using ClassicIndexSchemaFactory to use custom `schema.xml` file.
- `schema_source` - cookbook_file source for custom `schema.xml` file.
- `schema_cookbook` - cookbook_file cookbook for custom `schema.xml` file.

## License and Authors

Author:: Earth U (<iskitingbords@gmail.com>)


CHANGELOG.md
# app_solr CHANGELOG

This file is used to list changes made in each version of the app_solr cookbook.

## 2.0.0 - 2022-11-
### Changed
- Now supports only Chef 17.x and above

### Added
- Support for Solr 5.x
- Option to not set user ulimits

### Removed
- Non-user-interactive resource properties

## 1.1.0 - 2021-11-15
### Added
- Can now install Solr 6.6.6

### Changed
- Restart the service if custom config is installed

## 1.0.0 - 2021-11-08
### Added
- Initial release of the cookbook


test/cookbooks/test/metadata.rb
name             'test'
maintainer       'Earth U'
maintainer_email 'iskitingbords@gmail.com'
license          'Apache-2.0'
description      'Installs/Configures test'
version          '0.1.0'

depends 'app_solr'


test/cookbooks/test/Policyfile.rb
# Policyfile.rb - Describe how you want Chef Infra Client to build your system.
#
# For more information on the Policyfile feature, visit
# https://docs.chef.io/policyfile/

# A name that describes what the system you're building with Chef does.
name 'test'

# Where to find external cookbooks:
default_source :supermarket

# run_list: chef-client will run these recipes in the order specified.
run_list 'test::default'

# Specify a custom source for a single cookbook:
cookbook 'test', path: '.'


test/cookbooks/test/recipes/default.rb
#
# Cookbook:: test
# Recipe:: default
#
# Copyright:: 2021, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# So Inspec can use node attributes
ruby_block 'save node attribs' do
  block do
    ::File.write('/tmp/kitchen_chef_node.json', node.to_json)
  end
end

app_solr_standalone node['test']['version'] do
  set_ulimits   node['test']['set_ulimits']
  solr_host     node['test']['solr_host']
  solr_java_mem node['test']['solr_java_mem']
end

app_solr_core node['test']['core_name'] do
  use_custom_solrconfig true
  solrconfig_source     node['test']['core_src'][node['test']['version']]['solrconfig']
  solrconfig_cookbook   'app_solr'

  use_custom_schema true
  schema_source     node['test']['core_src'][node['test']['version']]['schema']
  schema_cookbook   'app_solr'
end


resources/core.rb
#
# Cookbook:: app_solr
# Resource:: standalone_core
#
# Copyright:: 2022, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'pathname'
cb = 'app_solr'

unified_mode true

# Create a core in standalone Solr

# Check status of local standlone Solr core:
#   curl 'http://localhost:8983/solr/admin/cores?action=STATUS&wt=json&core=corename'

# By default, this resource creates a core using version 7.7.3
# solrconfig.xml and schema.xml files. Custom files can be provided
# as properties.
# The cookbook also includes default solrconfig.xml and schema.xml files
# for each version: 5.5.5, 6.6.6, and 7.7.3.

property :name, String,
         description: 'Name of Solr core to create',
         name_property: true

property :extract_dir, String,
         description: 'Location of the extracted Solr installation files',
         default: '/opt'

property :solr_user, String,
         description: 'Name of Solr user',
         default: 'solr'

# Used only to provide a consistent :solr_home property
# between the app_solr_core and app_solr_standalone resources.
property :solr_dir, String,
         description: 'Main directory for Solr. '\
                      'Only used if :solr_home is relative.',
         default: '/var/solr'

property :solr_home, String,
         description: 'Location of Solr home. Can be relative to :solr_dir.',
         default: 'data'

property :use_custom_solrconfig, [true, false],
         description: 'Whether or not to use a custom solrconfig.xml. If '\
                      'true, provide values for :solrconfig_source and '\
                      ':solrconfig_cookbook.',
         default: false

property :solrconfig_source, String,
         description: 'Name of the cookbook_file source for a '\
                      'custom solrconfig.xml',
         default: 'solrconfig.xml'

property :solrconfig_cookbook, String,
         description: 'Cookbook of cookbook_file for custom solrconfig.xml',
         default: cb

property :use_custom_schema, [true, false],
         description: 'Whether or not to use a custom schema.xml. If '\
                      'true, provide values for :schema_source and '\
                      ':schema_cookbook. Useful if using the '\
                      '"ClassicIndexSchemaFactory" schemaFactory in :solrconfig ',
         default: false

property :schema_source, String,
         description: 'Name of the cookbook_file source for a '\
                      'custom schema.xml',
         default: 'schema.xml'

property :schema_cookbook, String,
         description: 'Cookbook of cookbook_file for custom schema.xml',
         default: cb

action_class do
  def form_solr_path(x)
    if x == ''
      new_resource.solr_dir
    else
      Pathname.new(x).absolute? ? x : "#{new_resource.solr_dir}/#{x}"
    end
  end
end

action :create do
  core_dir = "#{form_solr_path(new_resource.solr_home)}/#{new_resource.name}"

  solr_bin = "#{new_resource.extract_dir}/solr/bin/solr"
  execute "create_solr_core_#{new_resource.name}" do
    command "#{solr_bin} create -c #{new_resource.name}"
    user    new_resource.solr_user
    group   new_resource.solr_user
    not_if  { ::File.exist?("#{core_dir}/core.properties") }
  end

  cookbook_file "#{core_dir}/conf/solrconfig.xml" do
    source   new_resource.solrconfig_source
    cookbook new_resource.solrconfig_cookbook
    owner    new_resource.solr_user
    group    new_resource.solr_user
    mode     '0660'
    only_if  { new_resource.use_custom_solrconfig }
  end

  cookbook_file "#{core_dir}/conf/schema.xml" do
    source   new_resource.schema_source
    cookbook new_resource.schema_cookbook
    owner    new_resource.solr_user
    group    new_resource.solr_user
    mode     '0660'
    only_if  { new_resource.use_custom_schema }
  end

  service 'solr' do
    action  :restart
    only_if { new_resource.use_custom_solrconfig || new_resource.use_custom_schema }
  end
end


resources/standalone.rb
#
# Cookbook:: app_solr
# Resource:: standalone
#
# Copyright:: 2022, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'pathname'
cb = 'app_solr'

unified_mode true

# Installation of standalone Solr 7.7.x as per this guide:
# https://solr.apache.org/guide/7_7/taking-solr-to-production.html

# Installation of standalone Solr 6.6.x as per this guide:
# https://solr.apache.org/guide/6_6/taking-solr-to-production.html

# Installation of standalone Solr 5.5.x as per this guide:
# http://archive.apache.org/dist/lucene/solr/ref-guide/apache-solr-ref-guide-5.5.pdf

# Check status of local standalone Solr:
#   curl http://localhost:8983/solr/admin/info/system?wt=json

property :version, String,
         description: 'Solr version to install. Only supports 5.x, 6.x, 7.x.',
         name_property: true

property :apt_packages, Array,
         description: 'Apt packages to install (e.g. openjdk packages). '\
                      'Default: openjdk-11 packages for Solr 7.x and '\
                      'openjdk-8 packages for Solr 6.x and below.'

property :extract_dir, String,
         description: 'Location to which the tarball will be extracted',
         default: '/opt'

property :solr_user, String,
         description: 'Name of user that runs Solr. This user is '\
                      'auto-created by the install script',
         default: 'solr'

property :set_ulimits, [true, false],
         description: 'If true, ulimits for :solr_user will be '\
                      'maximized to 65535',
         default: true

property :solr_dir, String,
         description: 'Main directory for Solr',
         default: '/var/solr'

property :force_install, [true, false],
         description: 'Set to true when upgrading Solr',
         default: false

property :solr_home, String,
         description: 'A setting in solr.in.sh',
         default: 'data'

property :solr_pid_dir, String,
         description: 'A setting in solr.in.sh',
         default: ''

property :solr_logs_dir, String,
         description: 'A setting in solr.in.sh',
         default: 'logs'

property :solr_host, String,
         description: 'A setting in solr.in.sh',
         default: '0.0.0.0'

property :solr_port, [Integer, String],
         description: 'A setting in solr.in.sh',
         default: '8983'

property :solr_java_mem, String,
         description: 'A setting in solr.in.sh',
         default: '-Xms1536m -Xmx1536m'

property :log4j_props, String,
         description: 'A setting in solr.in.sh',
         default: 'log4j2.xml'

property :other_props, Hash,
         description: 'Additional settings in solr.in.sh',
         default: {}

action_class do
  def form_solr_path(x)
    if x == ''
      new_resource.solr_dir
    else
      Pathname.new(x).absolute? ? x : "#{new_resource.solr_dir}/#{x}"
    end
  end
end

action :install do
  tmp = Chef::Config[:file_cache_path]

  # If major version is not here, it's not supported
  verprops = {
    '7' => {
      packages: %w(openjdk-11-jdk-headless openjdk-11-jre-headless),
      opts:     ' -n',
      conf_src: 'solr.in.sh_7.7.3.erb',
      action:   :start,
    },
    '6' => {
      packages: %w(openjdk-8-jdk-headless openjdk-8-jre-headless),
      opts:     ' -n',
      conf_src: 'solr.in.sh_6.6.6.erb',
      action:   :start,
    },
    '5' => {
      packages: %w(openjdk-8-jdk-headless openjdk-8-jre-headless),
      opts:     '',
      conf_src: 'solr.in.sh_5.5.5.erb',
      action:   :restart,
    },
  }
  mv            = new_resource.version.split('.')[0]
  major_version = verprops[mv]

  apt_update
  major_version[:packages].each do |p|
    package p
  end

  source_url = format(
                 'https://archive.apache.org/dist/lucene/solr/%s/solr-%s.tgz',
                 new_resource.version,
                 new_resource.version
               )
  tarball_name   = ::File.basename(source_url)
  install_script = 'bin/install_solr_service.sh'

  remote_file "#{tmp}/#{tarball_name}" do
    source source_url
  end

  tarball_basename = tarball_name.chomp(::File.extname(tarball_name))
  strip_comp       = install_script.split('/').length
  execute 'extract_solr_install_script' do
    cwd     tmp
    command "tar xzf #{tarball_name} #{tarball_basename}/#{install_script} "\
            "--strip-components=#{strip_comp}"
  end

  extracted_script = ::File.basename(install_script)
  solr_bin         = "#{new_resource.extract_dir}/solr/bin/solr"
  install_opts     = "-i #{new_resource.extract_dir} "\
                     "-d #{new_resource.solr_dir} "\
                     "-u #{new_resource.solr_user}"
  if new_resource.force_install
    install_opts << ' -f'
  end
  install_opts << major_version[:opts]
  execute 'install_solr' do
    cwd     tmp
    command "bash #{extracted_script} #{tarball_name} #{install_opts}"
    not_if  { ::File.exist?(solr_bin) && !new_resource.force_install }
  end

  if new_resource.set_ulimits
    user_ulimit new_resource.solr_user do
      filehandle_hard_limit 65535
      filehandle_soft_limit 65535
      process_hard_limit    65535
      process_soft_limit    65535
    end
  end

  template '/etc/default/solr.in.sh' do
    cookbook  cb
    source    major_version[:conf_src]
    owner     'root'
    group     new_resource.solr_user
    mode      '0640'
    variables(
      solr_home:     form_solr_path(new_resource.solr_home),
      solr_pid_dir:  form_solr_path(new_resource.solr_pid_dir),
      solr_logs_dir: form_solr_path(new_resource.solr_logs_dir),
      log4j_props:   form_solr_path(new_resource.log4j_props),
      solr_host:     new_resource.solr_host,
      solr_port:     new_resource.solr_port,
      solr_java_mem: new_resource.solr_java_mem,
      other_props:   new_resource.other_props
    )
  end

  service 'solr' do
    action major_version[:action]
  end
end
